ufl/algorithms/apply_restrictions.py

```diff
@@ -9,178 +9,39 @@
-    terminal = _missing_rule
-
-    multi_index = _ignore_restriction
-    label = _ignore_restriction
-
-    # Default: Literals should ignore restriction
-    constant_value = _ignore_restriction
-    constant = _ignore_restriction
-
-    # Even arguments with continuous elements such as Lagrange must be
-    # restricted to associate with the right part of the element
-    # matrix
-    argument = _require_restriction
-
-    # Defaults for geometric quantities
-    geometric_cell_quantity = _require_restriction
-    geometric_facet_quantity = _require_restriction
-
-    # Only a few geometric quantities are independent on the restriction:
-    facet_coordinate = _ignore_restriction
-    quadrature_weight = _ignore_restriction
-
-    # Assuming homogeoneous mesh
-    reference_cell_volume = _ignore_restriction
-    reference_facet_volume = _ignore_restriction
-
-    def coefficient(self, o):
-        """Restrict a coefficient.
-
-        Allow coefficients to be unrestricted (apply default if so) if
-        the values are fully continuous across the facet.
-        """
-        if o.ufl_element() in H1:
-            # If the coefficient _value_ is _fully_ continuous
-            # It must still be computed from one of the sides, we just don't care which
-            return self._default_restricted(o)
-        else:
-            return self._require_restriction(o)
-
-    def facet_normal(self, o):
-        """Restrict a facet_normal."""
-        D = extract_unique_domain(o)
-        e = D.ufl_coordinate_element()
-        gd = D.geometric_dimension()
-        td = D.topological_dimension()
-
-        if e.embedded_superdegree <= 1 and e in H1 and gd == td:
-            # For meshes with a continuous linear non-manifold
-            # coordinate field, the facet normal from side - points in
-            # the opposite direction of the one from side +.  We must
-            # still require a side to be chosen by the user but
-            # rewrite n- -> n+.  This is an optimization, possibly
-            # premature, however it's more difficult to do at a later
-            # stage.
-            return self._opposite(o)
-        else:
-            # For other meshes, we require a side to be
-            # chosen by the user and respect that
-            return self._require_restriction(o)
-
-
-def apply_restrictions(expression):
+def apply_restrictions(expression: ApplyRestrictions):
     """Propagate restriction nodes to wrap differential terminals directly."""
-    integral_types = [
-        k for k in integral_type_to_measure_name.keys() if k.startswith("interior_facet")
-    ]
-    rules = RestrictionPropagator()
-    return map_integrand_dags(rules, expression, only_integral_type=integral_types)
+    warnings.warn(
+        "apply_restrictions(expression) is deprecated and will be removed "
+        "after December 2024. Please, use expression.apply_restrictions() instead",
+        FutureWarning,
+    )
```

Comment:
Do we want to deprecate this function and encourage use of `expression.apply_restrictions()` instead? Or do we want to keep it?